### PR TITLE
Implement memory initialization state copy functionality 

### LINKED ISSUE #3347: Checking memory initialization in presence of `copy` and `copy_nonoverlapping` produces false positives
Trying to verify the following code with memory initialization checks enabled (`-Z uninit-checks`):

```rust
#[derive(kani::Arbitrary)]
struct S(u32, u8);

#[kani::proof]
unsafe fn main() {
    let mut a: S = kani::any();
    let b: S = kani::any();
    let a_ptr = &mut a as *mut S;
    std::ptr::write_unaligned(a_ptr, b);
}
```

yield the following error:

```
Failed Checks: Undefined Behavior: Reading from an uninitialized pointer of type `*const u8`
 File: ".../.rustup/toolchains/nightly-2024-07-15-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/intrinsics.rs", line 2958, in std::intrinsics::copy_nonoverlapping::<u8>
```

with Kani version: 0.53.0

I expected to see this happen:
```
VERIFICATION:- SUCCESSFUL
```

According to the Rust documentaiton, both `copy` and `copy_nonoverlapping` are [untyped](https://doc.rust-lang.org/nightly/std/ptr/fn.copy.html):

> The copy is “untyped” in the sense that data may be uninitialized or otherwise violate the requirements of T. The initialization state is preserved exactly.

However, Kani treats a copy as a read followed by a write, which triggers memory initialization checks while it shouldn't. This becomes evident since the source code of `write_unaligned` copies the data byte-by-byte. 
```rust
pub const unsafe fn write_unaligned<T>(dst: *mut T, src: T) {
    // SAFETY: the caller must guarantee that `dst` is valid for writes.
    // `dst` cannot overlap `src` because the caller has mutable access
    // to `dst` while `src` is owned by this function.
    unsafe {
        copy_nonoverlapping(addr_of!(src) as *const u8, dst as *mut u8, mem::size_of::<T>());
        // We are calling the intrinsic directly to avoid function calls in the generated code.
        intrinsics::forget(src);
    }
}
```

### PR DESCRIPTION
**Pull request by artemagvanian**
This PR adds support of copying memory initialization state without checks in-between. Every time a copy is performed, the tracked byte is non-deterministically switched.

Resolves #3347

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/kani/src/mem_init.rs [hunk 1]
```diff
@@ -89,6 +89,33 @@ impl MemoryInitializationState {
         }
     }
 
+    /// Copy memory initialization state by non-deterministically switching the tracked object and
+    /// adjusting the tracked offset.
+    pub fn copy<const LAYOUT_SIZE: usize>(
+        &mut self,
+        from_ptr: *const u8,
+        to_ptr: *const u8,
+        num_elts: usize,
```
[the remark this answers is not in the document]
Comment (celinval) [reply library/kani/src/mem_init.rs:98]: Another option would be to change these interfaces to take `T` and use `size_of::<T>::()` which is cleaner.

**Review by celinval — COMMENTED** [review]
As discussed offline, copying can trigger a delayed UB in the presence of pointer casting from types with different padding bytes location. Can you please mitigate this case here though? We could encode copy operations as unsupported. Or we could also keep a global variable that we set whenever an incompatible cast is performed. Whenever we run `copy` or `copy_nonoverlapping`, we can assert that global variable hasn't been set.

**Comment by artemagvanian** [comment]
I am considering waiting until we make some progress with supporting delayed UB triggered by pointer casting with different padding and revisiting this PR then

File: library/kani/src/mem_init.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (artemagvanian) [reply library/kani/src/mem_init.rs:100]: Having the types to be `*const u8` is consistent with the other methods. They are not exposed in the public interface, but perhaps we should change all of them in another PR.

**Comment by artemagvanian** [comment]
> Can you please add a test that has a more complex CFG / Call graph combination? Maybe the reference comes from a different function than the one that creates the raw pointer and does the copy? Some branches might be nice too. :)

That's a good idea. I added a test with a more complex call and control flow structure. I also added a couple of tests checking that this interacts well with delayed UB detection.

**Pull request merged**